AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,352 @@
+  
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didPlayToEnd:) name:AVPlayerItemDidPlayToEndTimeNotification object:nil];
+  
+  return self;
+}
+
+// FIXME: Adopt interfaceStateDidChange API
+- (void)setInterfaceState:(ASInterfaceState)newState
+{
+  [super setInterfaceState:newState];
+  
+  if (!(newState & ASInterfaceStateVisible)) {
+    [self pause];
+    [(UIActivityIndicatorView *)_spinner.view stopAnimating];
+    [_spinner removeFromSupernode];
+  } else {
+    if (_shouldBePlaying) {
+      [self play];
+    }
+  }
+
+  if (newState & ASInterfaceStateVisible) {
+    [self displayDidFinish];
+  }
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+{
+  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusReadyToPlay) {
+    if ([self.subnodes containsObject:_spinner]) {
+      [_spinner removeFromSupernode];
+      _spinner = nil;
+    }
+  }
+  
+  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusFailed) {
+    
+  }
+}
+
+- (void)didPlayToEnd:(NSNotification *)notification
+{
+  if (ASObjectIsEqual([[notification object] asset], _asset)) {
+    [_delegate videoDidReachEnd:self];
+    [_player seekToTime:CMTimeMakeWithSeconds(0, 1)];
+    
+    if (_autorepeat) {
+      [self play];
+    } else {
+      [self pause];
+    }
+  }
+}
+
+- (void)layout
+{
+  [super layout];
+  _playerNode.frame = self.bounds;
+  
+  CGFloat horizontalDiff = (self.bounds.size.width - _playButton.bounds.size.width)/2;
+  CGFloat verticalDiff = (self.bounds.size.height - _playButton.bounds.size.height)/2;
+  _playButton.hitTestSlop = UIEdgeInsetsMake(-verticalDiff, -horizontalDiff, -verticalDiff, -horizontalDiff);
+  
+  _spinner.frame = _playButton.frame;
+}
+
+- (void)tapped
+{
+  if (_shouldBePlaying) {
+    [self pause];
+  } else {
+    [self play];
+  }
+}
+
+- (instancetype)initWithViewBlock:(ASDisplayNodeViewBlock)viewBlock didLoadBlock:(ASDisplayNodeDidLoadBlock)didLoadBlock
+{
+  ASDisplayNodeAssertNotSupported();
+  return nil;
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+
+  @try {
+    [_currentItem removeObserver:self forKeyPath:NSStringFromSelector(@selector(status))];
+  }
+  @catch (NSException * __unused exception) {
+    NSLog(@"unnecessary removal in fetch data");
+  }
+
+  {
+    ASDN::MutexLocker l(_lock);
+    _currentItem = [[AVPlayerItem alloc] initWithAsset:_asset];
+    [_currentItem addObserver:self forKeyPath:NSStringFromSelector(@selector(status)) options:NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew context:NULL];
+
+    if (_player) {
+      [_player replaceCurrentItemWithPlayerItem:_currentItem];
+    } else {
+      _player = [[AVPlayer alloc] initWithPlayerItem:_currentItem];
+    }
+  }
+}
+
+// FIXME: Adopt interfaceStateDidChange API
+- (void)displayDidFinish
```

Comment:
I think this can be visibilityDidChange: - specifically should not be this method, because display may occur more than two screens away due to pre-rendering